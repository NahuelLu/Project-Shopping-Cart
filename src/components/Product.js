import '../styles/Product.css'
import { useState} from 'react'
import { useShopContext } from './ShopContext'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faCartPlus, faMinus, faPlus } from '@fortawesome/free-solid-svg-icons'
import PropTypes from 'prop-types'

const alreadyExistsBookInCart =  (book, cart) => cart.some(item => item.id===book.id)
const addAmountItem = (itemCart, amountToAdd) => {return    {...itemCart,amount:itemCart.amount + amountToAdd}}
const updateCartWithBookSelected = (book, cart,amountItem) =>  cart.map( itemCart => (itemCart.id === book.id)?addAmountItem(itemCart,amountItem):itemCart) 
const addNewBookToCart = (cart, book,amountItem) => {
    const {title,imageLinks} = book.volumeInfo
    const {amount} = book.saleInfo.listPrice
    const newBook = {title:title,id:book.id,img:imageLinks.thumbnail,price:amount,amount:amountItem}
    return cart.concat([newBook])
}
const updateCart = (cart,book,amountItem)=> {
   return alreadyExistsBookInCart(book, cart) ? updateCartWithBookSelected(book,cart,amountItem): addNewBookToCart(cart,book,amountItem)
}
const bookHasRequiredDataToShowIt = (book)=> book.volumeInfo.imageLinks && book.volumeInfo.imageLinks.thumbnail && book.volumeInfo.description && book.saleInfo.listPrice





const Product = ({book})=>{
    const [buyStatus, setBuyStatus] = useState(false)
    const [amountBook, setAmountBook] = useState(0)
    const {setItemsCart} = useShopContext()

    const changeBuyState = ()=>{
        buyStatus? setBuyStatus(false) : setBuyStatus(true) 
    }
    const handlerAmountBook = (e)=>{
        setAmountBook(Number(e.currentTarget.value))
    }
    const upAmountBook = (e)=> {
        e.preventDefault();
        setAmountBook(prevAmount => prevAmount + 1)
    }
    const downAmountBook = (e)=> {
        e.preventDefault();
        setAmountBook(prevAmount => prevAmount > 0?prevAmount-1:0)
    }
    const handleSubmit = (e)=>{
        e.preventDefault()
        setItemsCart(prevItemsCart => updateCart(prevItemsCart,book,amountBook) )
        setBuyStatus(false)
    }
    
    return(
        <>
        {bookHasRequiredDataToShowIt(book) &&
            <div className='book-container'>
                <div>
                    <img src={book.volumeInfo.imageLinks.thumbnail} alt="Book thumbnail" />
                </div>
                <div>${book.saleInfo.listPrice.amount}</div>
                <button onClick={changeBuyState} aria-label='buy'>
                    <FontAwesomeIcon size='2x'icon={buyStatus?faMinus:faPlus} />
                </button>
                {buyStatus && 
                    <form className='buy-container' onSubmit={handleSubmit}>
                        <label htmlFor='itemAmount'>Amount</label>
                        <input onChange={handlerAmountBook} type='number'id='itemAmount' name='itemAmount' value={amountBook}></input>
                        <div className='buttons-container'>
                            <button onClick={downAmountBook} aria-label='downAmount'>
                                <FontAwesomeIcon icon={faMinus} />
                            </button>
                            <button>
                            <FontAwesomeIcon size='2x'icon={faCartPlus} />
                            </button>
                            <button onClick={upAmountBook} aria-label='upAmount'>
                                <FontAwesomeIcon icon={faPlus} />
                            </button>
                        </div>
                    </form>
                }
            </div>
        }
        </>
    )
}
Product.propTypes ={
    book : PropTypes.object
}

export default Product