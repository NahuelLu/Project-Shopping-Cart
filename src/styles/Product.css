.book-container{
    display: flex;
    flex-direction: column;
    align-items: center;
    width: 30%;
    justify-content: center;
}
.buy-container{
    display: flex;
    flex-direction: column;
    align-items: center;
}