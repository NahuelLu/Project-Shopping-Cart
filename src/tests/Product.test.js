import { act, render, screen } from "@testing-library/react"
import Product from "../components/Product"
import userEvent from '@testing-library/user-event'

const goodBook = {
    "id": "hvTAEAAAQBAJ",
    "volumeInfo": {
      "title": "No quiero perderme nada",
      "imageLinks": {
        "thumbnail": "http://books.google.com/books/content?id=hvTAEAAAQBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api"
      },
      "description": "Hola"
    },
    "saleInfo": {
      "listPrice": {
        "amount": 1469.99,
      },
    },
}
jest.mock("../components/ShopContext", () => {
    return {
        useShopContext: ()=> ({
            setItemsCart: ()=>{}
        })
    }
  }
);
jest.mock("@fortawesome/react-fontawesome", ()=>{
    return {
        FontAwesomeIcon: ({icon})=> <div>{icon}</div>
    }
})
jest.mock("@fortawesome/free-solid-svg-icons",()=>{
    return{
        faCartPlus: "faCartPlusIcon",
        faMinus : "faMinusIcon",
        faPlus : "faPlusIcon"
    }
})
jest.mock('react-router-dom',()=>{
    return {
      Link : ({to})=> <a>Url:{to}</a>
    }
})
describe("Tests for Product component",()=>{
    
    test("Render successfully book",()=>{
        const {container}=render(<Product book={goodBook} />)
        expect(container).toMatchSnapshot()
    })
    test("Don't render book because doesn't has imageLinks",()=>{
        const book = {
            "id": "hvTAEAAAQBAJ",
            "volumeInfo": {
              "title": "No quiero perderme nada",
              "description": "Hola"
            },
            "saleInfo": {
              "listPrice": {
                "amount": 1469.99,
              },
            },
        }
        const {container} = render(<Product book={book} />)
        expect(book).not.toHaveProperty('volumeInfo.imageLinks.thumbnail');
        expect(container).toMatchSnapshot()
    })
    test("Don't render book because doesn't have a description",()=>{
        const book = {
            "id": "hvTAEAAAQBAJ",
            "volumeInfo": {
              "title": "No quiero perderme nada",
              "imageLinks": {
                "thumbnail": "http://books.google.com/books/content?id=hvTAEAAAQBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api"
              }
            },
            "saleInfo": {
              "listPrice": {
                "amount": 1469.99,
              },
            },
        }
        const {container} = render(<Product book={book} />)
        expect(book).not.toHaveProperty('volumeInfo.description');
        expect(container).toMatchSnapshot()
    })
    test("Don't render book because doesn't have a price",()=>{
        const book = {
            "id": "hvTAEAAAQBAJ",
            "volumeInfo": {
              "title": "No quiero perderme nada",
              "imageLinks": {
                "thumbnail": "http://books.google.com/books/content?id=hvTAEAAAQBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api"
              },
              "description": "Hola"
            },
            "saleInfo": {
            },
        }
        const {container} = render(<Product book={book} />)
        expect(book).not.toHaveProperty('saleInfo.listPrice');
        expect(container).toMatchSnapshot()
    })
    test("Render buy UI when click button",async()=>{
        const {container}=render(<Product book={goodBook} />)
        const user = userEvent.setup()
        const button = screen.getByRole("button",{name:'buy'})
        await act(async  ()=>{
            await user.click(button)
        })
        expect(container).toMatchSnapshot()
    })
    test("Up 1 to input amount",async()=>{
        const user = userEvent.setup()
        const {container}=render(<Product book={goodBook} />)
        const button = screen.getByRole("button",{name:'buy'})
        await act(async  ()=>{
            await user.click(button)
        })
        const btnUp = screen.getByRole("button",{name:"upAmount"})
        await act(async  ()=>{
            await user.click(btnUp)
        })
        const amount = container.querySelector("#itemAmount").value
        expect(amount).toBe("1")
        expect(container).toMatchSnapshot()
    })
    test("Up 1 then down 1 amount is 0",async()=>{
        const user = userEvent.setup()
        const {container}=render(<Product book={goodBook} />)
        const button = screen.getByRole("button",{name:'buy'})
        await act(async  ()=>{
            await user.click(button)
        })
        const btnUp = screen.getByRole("button",{name:"upAmount"})
        const btnDown = screen.getByRole("button",{name:"downAmount"})
        await act(async  ()=>{
            await user.click(btnUp)
            await user.click(btnDown)
        })
        const amount = container.querySelector("#itemAmount").value
        expect(amount).toBe("0")
    })
    
})