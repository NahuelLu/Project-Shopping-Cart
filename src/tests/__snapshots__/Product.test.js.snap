// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for Product component Don't render book because doesn't has imageLinks 1`] = `<div />`;

exports[`Tests for Product component Don't render book because doesn't have a description 1`] = `<div />`;

exports[`Tests for Product component Don't render book because doesn't have a price 1`] = `<div />`;

exports[`Tests for Product component Render buy UI when click button 1`] = `
<div>
  <div
    class="book-container"
  >
    <div>
      <img
        alt="Book thumbnail"
        src="http://books.google.com/books/content?id=hvTAEAAAQBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api"
      />
    </div>
    <div>
      $
      1469.99
    </div>
    <button
      aria-label="buy"
    >
      <div>
        faMinusIcon
      </div>
    </button>
    <form
      class="buy-container"
    >
      <label
        for="itemAmount"
      >
        Amount
      </label>
      <input
        id="itemAmount"
        name="itemAmount"
        type="number"
        value="0"
      />
      <div
        class="buttons-container"
      >
        <button
          aria-label="downAmount"
        >
          <div>
            faMinusIcon
          </div>
        </button>
        <button>
          <div>
            faCartPlusIcon
          </div>
        </button>
        <button
          aria-label="upAmount"
        >
          <div>
            faPlusIcon
          </div>
        </button>
      </div>
    </form>
  </div>
</div>
`;

exports[`Tests for Product component Render successfully book 1`] = `
<div>
  <div
    class="book-container"
  >
    <div>
      <img
        alt="Book thumbnail"
        src="http://books.google.com/books/content?id=hvTAEAAAQBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api"
      />
    </div>
    <div>
      $
      1469.99
    </div>
    <button
      aria-label="buy"
    >
      <div>
        faPlusIcon
      </div>
    </button>
  </div>
</div>
`;

exports[`Tests for Product component Up 1 to input amount 1`] = `
<div>
  <div
    class="book-container"
  >
    <div>
      <img
        alt="Book thumbnail"
        src="http://books.google.com/books/content?id=hvTAEAAAQBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api"
      />
    </div>
    <div>
      $
      1469.99
    </div>
    <button
      aria-label="buy"
    >
      <div>
        faMinusIcon
      </div>
    </button>
    <form
      class="buy-container"
    >
      <label
        for="itemAmount"
      >
        Amount
      </label>
      <input
        id="itemAmount"
        name="itemAmount"
        type="number"
        value="1"
      />
      <div
        class="buttons-container"
      >
        <button
          aria-label="downAmount"
        >
          <div>
            faMinusIcon
          </div>
        </button>
        <button>
          <div>
            faCartPlusIcon
          </div>
        </button>
        <button
          aria-label="upAmount"
        >
          <div>
            faPlusIcon
          </div>
        </button>
      </div>
    </form>
  </div>
</div>
`;
